hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/MetricsProxyEndpoint.java

```diff
@@ -0,0 +1,117 @@
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+import static org.apache.hadoop.ozone.recon.ReconConstants.PROMETHEUS_INSTANT_QUERY_API;
+
+/**
+ * Endpoint to fetch metrics data from Prometheus HTTP endpoint.
+ */
+@Path("/metrics")
+public class MetricsProxyEndpoint {
+
+  private MetricsServiceProvider metricsServiceProvider;
+
+  @Inject
+  public MetricsProxyEndpoint(
+      MetricsServiceProviderFactory metricsServiceProviderFactory) {
+    this.metricsServiceProvider =
+        metricsServiceProviderFactory.getMetricsServiceProvider();
+  }
+
+  /**
+   * Return a response from the configured metrics endpoint.
+   */
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/{api}")
+  public void getMetricsResponse(
+      @DefaultValue (PROMETHEUS_INSTANT_QUERY_API) @PathParam("api") String api,
+      @Context UriInfo uriInfo,
+      @Context HttpServletResponse httpServletResponse
+  ) throws Exception {
+    if (metricsServiceProvider != null) {
+      HttpURLConnection connection = metricsServiceProvider.getMetricsResponse(
+          api, uriInfo.getRequestUri().getQuery());
+      InputStream inputStream;
+      if (Response.Status.fromStatusCode(connection.getResponseCode())
+          .getFamily() == Response.Status.Family.SUCCESSFUL) {
+        inputStream = connection.getInputStream();
+      } else {
+        // Throw a bad gateway error if HttpResponseCode is not 2xx
+        httpServletResponse.setStatus(HttpServletResponse.SC_BAD_GATEWAY);
+        inputStream = connection.getErrorStream();
+      }
+      try (
+          OutputStream outputStream =
+              httpServletResponse.getOutputStream();
+          ReadableByteChannel inputChannel =
+              Channels.newChannel(inputStream);
+          WritableByteChannel outputChannel =
+              Channels.newChannel(outputStream)
+        ) {
+        final ByteBuffer buffer = ByteBuffer.allocateDirect(16 * 1024);
+
+        while(inputChannel.read(buffer) != -1) {
+          buffer.flip();
+          outputChannel.write(buffer);
+          buffer.compact();
+        }
+
+        buffer.flip();
+
+        while(buffer.hasRemaining()) {
+          outputChannel.write(buffer);
+        }
+      } finally {
+        inputStream.close();
+      }
+    } else {
+      // Throw a Bad Gateway Error
+      httpServletResponse.sendError(HttpServletResponse.SC_BAD_GATEWAY,
+          "Metrics endpoint is not configured. Configure " +
+              PrometheusServiceProviderImpl.getEndpointConfigKey() + " and " +
```

Comment:
Nit. getEndpointConfigKey() can be a method on the MetricsServiceProvider interface.

Reply:
@vivekratnavel Makes sense, thanks!